gcp/workers/alias/upstream_computation.py

```diff
@@ -76,30 +79,101 @@ def _update_group(upstream_group, upstream_ids: list):
   upstream_group.put()
 
 
+def compute_upstream_hierarchy(target_upstream_group: osv.UpstreamGroup,
+                               all_upstream_groups: ndb.Query):
+  """Computes all upstream vulnerabilities for the given bug ID.
+  The returned list contains all of the bug IDs that are upstream of the
+  target bug ID, including transitive upstreams in a map hierarchy.
+  upstream_group:
+        { db_id: bug id
+          upstream_ids: str[bug_ids]
+          last_modified_date: date
+          upstream_hierarchy: JSON str
+          }
+  """
+  visited = set()
+  upstream_map = {}
+  to_visit = set([target_upstream_group.db_id])
+
+  while to_visit:
+    bug_id = to_visit.pop()
+    if bug_id in visited:
+      continue
+    visited.add(bug_id)
+    upstream_group = all_upstream_groups.filter(
+        osv.UpstreamGroup.db_id == bug_id).get()
+    if upstream_group is None:
+      continue
+
+    upstreams = set(upstream_group.upstream_ids)
+    if not upstreams:
+      continue
+    for upstream in upstreams:
+      if upstream not in visited and upstream not in to_visit:
+        to_visit.add(upstream)
+      else:
+        if bug_id not in upstream_map:
+          upstream_map[bug_id] = set([upstream])
+        else:
+          upstream_map[bug_id].add(upstream)
+    upstream_map[bug_id] = upstreams
+    to_visit.update(upstreams - visited)
+
+  for k, v in upstream_map.items():
+    if k is target_upstream_group.db_id:
+      continue
+    upstream_map[target_upstream_group
+                 .db_id] = upstream_map[target_upstream_group.db_id] - v
+
+  if target_upstream_group.upstream_hierarchy == upstream_map:
+    return
+
+  if upstream_map:
+    target_upstream_group.upstream_hierarchy = json.dumps(
+        upstream_map, default=set_default)
+    target_upstream_group.last_modified = datetime.datetime.now()
```

Comment:
I don't think we'll see this happen, but in theory, there could be a race condition where `bug.upstream_raw` updates between when we update `upstream_group` and `upstream_hierarchy`. I think it's better to not update the upstream's `last_modified` time here, since we already updated it when we updated `upstream_group`. And that modified time is only used to see if upstream_group needs to change.